### PR TITLE
[MPS] Enable NDHWC+DHWIO fast path for Conv3d on channels_last_3d

### LINKED ISSUE #182823: [MPS] Conv3d dominates 3D training cost; 3D-vs-2D ratio wider than CUDA after #180335
### 🐛 Describe the bug

After #180335 fixed the 5D `batch_norm_backward` slowdown that #178492 reported, `Conv3d` becomes the dominant cost in 3D training on MPS. At the workload level, the MPS 3D-vs-2D speedup ratio at equal numel is wider on Apple Silicon than on CUDA ([reference numbers in this comment](https://github.com/pytorch/pytorch/issues/178492#issuecomment-4183074182)).

### Repro

```python
import time
import torch
import torch.nn as nn

assert torch.backends.mps.is_available()
device = torch.device("mps")
torch.manual_seed(0)

def time_ms(label, fn, warmup=5, steps=30):
    for _ in range(warmup): fn()
    torch.mps.synchronize()
    t0 = time.perf_counter()
    for _ in range(steps): fn()
    torch.mps.synchronize()
    ms = (time.perf_counter() - t0) / steps * 1000
    print(f"  {label:>32s}: {ms:7.2f} ms")
    return ms

# 3D ops on x=[4,32,64,64,64]; 2D ops on x=[4,32,4096,128] (equal numel)
N = 4
x3 = torch.randn(N, 32, 64, 64, 64, device=device, requires_grad=True)
x2 = torch.randn(N, 32, 4096, 128, device=device, requires_grad=True)
c3 = nn.Conv3d(32, 64, 3, padding=1).to(device)
c2 = nn.Conv2d(32, 64, 3, padding=1).to(device)
bn3 = nn.BatchNorm3d(64).to(device).train()
bn2 = nn.BatchNorm2d(64).to(device).train()
p3 = nn.AdaptiveAvgPool3d(1)
p2 = nn.AdaptiveAvgPool2d(1)
y3 = torch.randn(N, 64, 64, 64, 64, device=device, requires_grad=True)
y2 = torch.randn(N, 64, 4096, 128, device=device, requires_grad=True)

print("3D ops, x=[4,32,64,64,64]:")
time_ms("Conv3d 32->64 fwd+bwd",      lambda: c3(x3).sum().backward())
time_ms("BatchNorm3d C=64 fwd+bwd",   lambda: bn3(y3).sum().backward())
time_ms("AdaptiveAvgPool3d fwd+bwd",  lambda: p3(y3).sum().backward())

print("\n2D ops, x=[4,32,4096,128] (equal numel):")
time_ms("Conv2d 32->64 fwd+bwd",      lambda: c2(x2).sum().backward())
time_ms("BatchNorm2d C=64 fwd+bwd",   lambda: bn2(y2).sum().backward())
time_ms("AdaptiveAvgPool2d fwd+bwd",  lambda: p2(y2).sum().backward())
```

### Observed (Apple Silicon M5, current `main` b16f9063b2)

```
3D ops, x=[4,32,64,64,64]:
       Conv3d 32->64 fwd+bwd:  381.32 ms
    BatchNorm3d C=64 fwd+bwd:   26.40 ms
   AdaptiveAvgPool3d fwd+bwd:   25.85 ms

2D ops, x=[4,32,4096,128] (equal numel):
       Conv2d 32->64 fwd+bwd:  213.73 ms
    BatchNorm2d C=64 fwd+bwd:   51.45 ms
   AdaptiveAvgPool2d fwd+bwd:   46.86 ms
```

`Conv3d 32→64 fwd+bwd` is `1.78x` of `Conv2d 32→64`, while Conv3d does `3x` more multiplies per output (`3x3x3=27` vs `3x3=9`). Per multiply, Conv3d is faster than Conv2d here, so the bare op is reasonable. The asymmetry shows up at the workload level.

### Workload-level ratio

Running the OP repro from #178492 (3D shape `[4,1,64,64,64]`) and the modified 2D repro (shape `[4,1,512,512]`, equal numel), full training step time:

|  | M5 post #180335 | M3 Max pre-fix | CUDA T4 |
|---|---|---|---|
| MPS 3D / MPS 2D step | 4.38x | 9.4x | 2.76x |
| CPU 3D / CPU 2D step | 3.01x | n/a | n/a |

(M3 Max and CUDA T4 numbers from the [linked comment](https://github.com/pytorch/pytorch/issues/178492#issuecomment-4183074182).)

CPU's `3.01x` is the floor (kernel size ratio). CUDA hits `2.76x`, MPS post-#180335 hits `4.38x`. Roughly half of the pre-fix gap is gone; the rest is residual.

In OP's 3D step, `Conv3d` is `~89%` of the time:

| Op (fwd+bwd) | Time | % of 3D step |
|---|---|---|
| Conv3d 32→64 | 378 ms | ~66% |
| Conv3d 1→32 | 132 ms | ~23% |
| BN3d C=64 | 26 ms | ~5% |
| AdaptiveAvgPool3d | 26 ms | ~5% |
| ReLU | 18 ms | ~3% |
| BN3d C=32 | 13 ms | ~2% |

### Notes

- `aten/src/ATen/native/mps/operations/Convolution.mm:122` disables channels_last for `is3DConv`. Whether channels_last_3d would be a win is open.
- Forward and backward go through `MPSGraph convolution3D...`, `convolution3DDataGradient...`, `convolution3DWeightsGradient...`. No custom Metal path.

### Versions

- PyTorch built from `b16f9063b2`
- macOS 26.1 (25B78)
- Apple Silicon M5, 32 GB

cc @malfet @kulinseth @DenisVieriu97 @hvaara

### PR DESCRIPTION
**Pull request by jhavukainen**
Fixes #182823.

This PR drops the `&& !is3DConv` guard so 3D convs with `channels_last_3d` input take the MPSGraph NDHWC data layout. It also adds an in-graph OIDHW<->DHWIO weight transpose gated by a beneficial-shape heuristic (Cin/groups >= 4, all kernel dims >= 2).  Forward, backward-input, and backward-weights all honor the layout-propagation contract after these changes.

For bf16/fp16 the underlying kernel for weight gradient computations is underperformant and it shows up especially with the more recent GPU generations. This is why the fp16/bf16 weight-grad computation is gated to take place in NCDHW+OIDHW until there is a kernel side fix to avoid regressing the perf.

Benchmarking this with M2 Max and M5 Max sweeping across some commonly encountered shapes using the CL3d now that it is respected for the computation

```
  ┌──────────────────────────────┬─────────────────────────────────────┬───────┬────────────┬─────────────┬────────────────┬────────────────┐
  │            Sweep             │                range                │ dtype │ M2 Max fwd │ M5 Max fwd  │ M2 Max fwd+bwd │ M5 Max fwd+bwd │
  ├──────────────────────────────┼─────────────────────────────────────┼───────┼────────────┼─────────────┼────────────────┼────────────────┤
  │ Spatial (Cin=32, Cout=64)    │ S ∈ {8, 16, 24, 32, 48, 64}         │ fp32  │ 1.30–1.43x │ 0.94–1.28x  │ 1.05–1.18x     │ 0.88–1.07x     │
  ├──────────────────────────────┼─────────────────────────────────────┼───────┼────────────┼─────────────┼────────────────┼────────────────┤
  │                              │                                     │ fp16  │ 1.10–1.25x │ 0.86–1.13x  │ 1.02–1.06x     │ 0.99–1.04x     │
  ├──────────────────────────────┼─────────────────────────────────────┼───────┼────────────┼─────────────┼────────────────┼────────────────┤
  │                              │                                     │ bf16  │ 1.09–1.25x │ 0.83–1.16x  │ 1.02–1.07x     │ 0.88–1.04x     │
  ├──────────────────────────────┼─────────────────────────────────────┼───────┼────────────┼─────────────┼────────────────┼────────────────┤
  │ Channel (S=32)               │ C=Cin=Cout ∈ {16, 32, 64, 128, 256} │ fp32  │ 1.04–1.38x │ 1.00–1.28x  │ 0.87–1.26x     │ 0.81–1.18x     │
  ├──────────────────────────────┼─────────────────────────────────────┼───────┼────────────┼─────────────┼────────────────┼────────────────┤
  │                              │                                     │ fp16  │ 1.02–1.32x │ 0.97–1.18x  │ 0.99–1.15x     │ 0.98–1.10x     │
  ├──────────────────────────────┼─────────────────────────────────────┼───────┼────────────┼─────────────┼────────────────┼────────────────┤
  │                              │                                     │ bf16  │ 1.04–1.35x │ 0.99–1.13x  │ 0.99–1.15x     │ 0.99–1.05x     │
  ├──────────────────────────────┼─────────────────────────────────────┼───────┼────────────┼─────────────┼────────────────┼────────────────┤
  │ First-layer (Cin=3, Cout=32) │ S ∈ {64, 96, 128}                   │ all   │ ~1.00x     │ ~1.00–1.02x │ 0.92–0.97x     │ 0.90–0.96x     │
  └──────────────────────────────┴─────────────────────────────────────┴───────┴────────────┴─────────────┴────────────────┴────────────────┘
```

So the story is somewhat mixed: Forward pass universally benefits from the change. The full fwd+bw pass sometimes benefits and sometimes not. The issue is more prevalent with the wider GPUs where the insufficiently exposed parallelism becomes an issue dropping the GPU util and with half types, while smaller GPUs are kept busy enough regardless to have a lesser impact. However once the faulty weight gradient kernel is tuned there should be a second uplift. This is the remaining gap mentioned in #182823 

The reasoning for accepting that downside is that now the conv3d op respects the channels last layout wishes from the caller by returning its outputs as channels last as well possibly helping with downstream usage. This did not use to be case. Also I tried to avoid baking in too much complexity on choosing when to respect the CL3d for the computation and when not to just to work around the current quirks of the underlying kernel.

One last note on the discrepancy of the NDHWC + DHWIO result in the follow-up comment of #182823: The test presented there does not account for pytorch weights being stored as OIDHW. So while the computation objectively will benefit from DHWIO used in conjunction with CL3d we will need to pay the cost of manipulating the weights as well as the gradients to them to align with their storage. This means that in reality the permutations will eat some of the perf benefits that the standalone test expected due to its optimistic setup.

### CODE REVIEW

**Comment by qflen** [comment]
Thanks for picking this up. The OIDHW storage point is correct... The standalone test sidestepped the host-side permute cost, so the 1.19x was optimistic and the real win lands lower.

Also forward universal plus bf16/fp16 weight grad gate makes sense to me as a layout respecting contract.

**Review by malfet — APPROVED** [review]

**Comment by jhavukainen** [comment]
Seems like an unrelated issue that's been fixed elsewhere. Rebasing the branch.
